nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,428 @@
+
+        condition: batch of blurred spectrograms
+        lenghts: length for every spectrogam in the batch
+        zs: latent noise inputs on the unit sphere (either this or ws or neither)
+        ws: latent noise inputs in the style space (either this or zs or neither)
+        noise: per-pixel indepentent gaussian noise
+        mixing: style mixing, usually True during training
+        normalize: normalize spectrogram range to ~[0, 1], True for normal use
+
+        For explanation of style mixing refer to [1]
+        For definititions of z, w [2]
+        [1] Karras et. al. - A Style-Based Generator Architecture for Generative Adversarial Networks, 2018 (https://arxiv.org/abs/1812.04948)
+        [2] Karras et. al. - Analyzing and Improving the Image Quality of StyleGAN, 2019 (https://arxiv.org/abs/1912.04958)
+        """
+        batch_size, *_, max_length = condition.shape
+
+        # generate noise
+        if zs is not None and ws is not None:
+            raise ValueError(
+                "Please specify either zs or ws or neither, but not both. It is not clear which one to use."
+            )
+
+        if zs is None:
+            zs = self.generate_zs(batch_size, mixing)
+        if ws is None:
+            ws = [self.G.style_mapping(z) for z in zs]
+        if noise is None:
+            noise = self.generate_noise(batch_size)
+
+        condition = rearrange(condition, "b c l -> b 1 c l")
+        # normalize if needed, mask and pad appropriately
+        if normalize:
+            condition = self.normalize_spectrograms(condition)
+        condition = self.pad_spectrogram(condition)
+        condition = mask(condition, lengths)
+
+        # the main call
+        enhanced_spectrograms = self.G(condition, lengths, ws, noise)
+
+        # denormalize if needed, mask and remove padding
+        if normalize:
+            enhanced_spectrograms = self.unnormalize_spectrograms(enhanced_spectrograms)
+        enhanced_spectrograms = mask(enhanced_spectrograms, lengths)
+        enhanced_spectrograms = enhanced_spectrograms[:, :, :, :max_length]
+        enhanced_spectrograms = rearrange(enhanced_spectrograms, "b 1 c l -> b c l")
+
+        return enhanced_spectrograms
+
+    def prepare_batch(self, batch):
+        if not isinstance(self.spectrogram_model, FastPitchModel):
+            raise NotImplementedError(
+                f"{type(self.spectrogram_model)} is not supported, please implement batch preparation for this model."
+            )
+
+        batch = process_batch(batch, self._train_dl.dataset.sup_data_types)
+        mels, mel_lens = self.spectrogram_model.preprocessor(input_signal=batch["audio"], length=batch["audio_lens"])
+
+        mels_pred, *_ = self.spectrogram_model(
+            text=batch["text"],
+            durs=None,
+            pitch=batch["pitch"],
+            speaker=batch.get("speaker"),
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
```

Comment:
We use the model to "resynthesize" input dataset, so the more ground-truth features the better.

`learn_alignment=True` means we FastPitch learns to compute alignment from input spectrogram and encoded text, which happens during training and exactly what we need here, because we need tts output spectrograms to have the same length as the input.

There is a bug here though: a model with `learn_alignment=False` is not handled correctly. `durs=None` always, so we most likely get exception somewhere down the line.